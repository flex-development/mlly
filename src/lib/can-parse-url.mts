/**
 * @file canParseUrl
 * @module mlly/lib/canParseUrl
 */

import type { ModuleId } from '@flex-development/mlly'

/**
 * Check if `input` can be parsed to a {@linkcode URL}.
 *
 * > ðŸ‘‰ **Note**: If `input` is relative, `base` is required. If `input` is
 * > absolute, `base` is ignored.
 *
 * @see {@linkcode ModuleId}
 *
 * @param {ModuleId} input
 *  The absolute or relative input URL to parse
 * @param {ModuleId | null | undefined} [base]
 *  Base URL to resolve against if `input` is not absolute
 * @return {boolean}
 *  `true` if `input` can be parsed to a `URL`
 */
function canParseUrl(
  input: ModuleId,
  base?: ModuleId | null | undefined
): boolean {
  try {
    new URL(input, base ?? undefined)
    return true
  } catch {
    return false
  }
}

export default canParseUrl
