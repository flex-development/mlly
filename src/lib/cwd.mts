/**
 * @file cwd
 * @module mlly/lib/cwd
 */

import pathe from '@flex-development/pathe'

/**
 * Get the URL of the current working directory.
 *
 * @return {URL}
 *  URL of current working directory
 */
function cwd(): URL {
  return pathe.pathToFileURL(pathe.cwd() + pathe.sep)
}

export default cwd
