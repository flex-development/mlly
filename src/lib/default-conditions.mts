/**
 * @file defaultConditions
 * @module mlly/lib/defaultConditions
 */

import type { Condition } from '@flex-development/pkg-types'

/**
 * Default conditions.
 *
 * @see {@linkcode Condition}
 * @see https://nodejs.org/api/packages.html#conditional-exports
 *
 * @const {Set<Condition>} defaultConditions
 */
const defaultConditions: Set<Condition> = new Set(['node', 'import'])

export default defaultConditions
