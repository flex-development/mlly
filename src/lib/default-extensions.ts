/**
 * @file defaultExtensions
 * @module mlly/lib/defaultExtensions
 */

import type { Ext } from '@flex-development/pathe'

/**
 * Default resolvable file extensions.
 *
 * @see {@linkcode Ext}
 *
 * @const {Set<Ext>} defaultExtensions
 */
const defaultExtensions: Set<Ext> = new Set([
  '.mjs',
  '.mts',
  '.cjs',
  '.cts',
  '.js',
  '.ts',
  '.jsx',
  '.tsx',
  '.css',
  '.json',
  '.node',
  '.wasm',
  '.d.mts',
  '.d.cts',
  '.d.ts'
])

export default defaultExtensions
