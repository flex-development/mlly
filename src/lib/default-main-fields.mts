/**
 * @file defaultMainFields
 * @module mlly/lib/defaultMainFields
 */

import type { MainField } from '@flex-development/mlly'

/**
 * Default main fields.
 *
 * @see {@linkcode MainField}
 *
 * @const {Set<MainField>} defaultMainFields
 */
const defaultMainFields: Set<MainField> = new Set(['main'])

export default defaultMainFields
