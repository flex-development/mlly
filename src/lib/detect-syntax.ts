/**
 * @file detectSyntax
 * @module mlly/lib/detectSyntax
 */

import hasCJSSyntax from './has-cjs-syntax'
import hasESMSyntax from './has-esm-syntax'

/**
 * Detects if `code` contains CommonJS syntax, ESM syntax, or a mixture of both.
 *
 * Ignores matches in comments.
 *
 * @see {@linkcode hasCJSSyntax}
 * @see {@linkcode hasESMSyntax}
 *
 * @param {string} code - Code to evaluate
 * @return {{ cjs: boolean; esm: boolean; mixed: boolean }} Detection result
 */
const detectSyntax = (
  code: string
): { cjs: boolean; esm: boolean; mixed: boolean } => {
  /**
   * CommonJS syntax check.
   *
   * @const {boolean} cjs
   */
  const cjs: boolean = hasCJSSyntax(code)

  /**
   * ESM syntax check.
   *
   * @const {boolean} esm
   */
  const esm: boolean = hasESMSyntax(code)

  return { cjs, esm, mixed: cjs && esm }
}

export default detectSyntax
