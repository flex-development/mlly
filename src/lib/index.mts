/**
 * @file Entry Point - Library
 * @module mlly/lib
 *
 * @todo `detectModuleSyntax` (esast integration)
 * @todo `findDynamicImports` (esast integration)
 * @todo `findExports` (esast integration)
 * @todo `findImports` (esast integration)
 * @todo `findRequires` (esast integration)
 * @todo `findSpecifiers` (esast integration)
 * @todo `findStaticImports` (esast integration)
 * @todo `resolveAliases` (esast integration)
 * @todo `resolveModules` (esast integration)
 */

export { default as canParseUrl } from '#lib/can-parse-url'
export { default as cwd } from '#lib/cwd'
export { default as defaultConditions } from '#lib/default-conditions'
export { default as defaultExtensions } from '#lib/default-extensions'
export { default as defaultMainFields } from '#lib/default-main-fields'
export { default as extensionFormatMap } from '#lib/extension-format-map'
export { default as formats } from '#lib/formats'
export { default as getSource } from '#lib/get-source'
export { default as isAbsoluteSpecifier } from '#lib/is-absolute-specifier'
export { default as isArrayIndex } from '#lib/is-array-index'
export { default as isBareSpecifier } from '#lib/is-bare-specifier'
export { default as isDirectory } from '#lib/is-directory'
export { default as isFile } from '#lib/is-file'
export { default as isImportsSubpath } from '#lib/is-imports-subpath'
export { default as isRelativeSpecifier } from '#lib/is-relative-specifier'
export { default as lookupPackageScope } from '#lib/lookup-package-scope'
export { default as patternKeyCompare } from '#lib/pattern-key-compare'
export { default as patternMatch } from '#lib/pattern-match'
export { default as readPackageJson } from '#lib/read-package-json'
export { default as resolveAlias } from '#lib/resolve-alias'
export { default as resolveModule } from '#lib/resolve-module'
export * from '#lib/resolver'
export * as resolver from '#lib/resolver'
export { default as root } from '#lib/root'
export { default as toRelativeSpecifier } from '#lib/to-relative-specifier'
