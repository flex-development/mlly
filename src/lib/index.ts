/**
 * @file Library
 * @module mlly/lib
 */

export { default as detectSyntax } from './detect-syntax'
export { default as extractStatements } from './extract-statements'
export { default as findDynamicImports } from './find-dynamic-imports'
export { default as findExports } from './find-exports'
export { default as findRequires } from './find-requires'
export { default as findStaticImports } from './find-static-imports'
export { default as hasCJSSyntax } from './has-cjs-syntax'
export { default as hasESMSyntax } from './has-esm-syntax'
export { default as resolveAlias } from './resolve-alias'
export { default as resolveAliases } from './resolve-aliases'
