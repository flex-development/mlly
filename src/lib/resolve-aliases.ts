/**
 * @file resolveAliases
 * @module mlly/lib/resolveAliases
 */

import type { AliasResolverOptions } from '#src/interfaces'
import * as pathe from 'pathe'
import { loadTsconfig, type Tsconfig } from 'tsconfig-paths/lib/tsconfig-loader'
import extractStatements from './extract-statements'
import resolveAlias from './resolve-alias'

/**
 * Batch path alias resolution options.
 *
 * @see {@link AliasResolverOptions}
 */
type Options = Omit<AliasResolverOptions, 'baseUrl' | 'paths' | 'url'>

/**
 * Resolves path aliases in `code`.
 *
 * @see [`resolveAlias`](./resolve-alias)
 *
 * @param {string} code - Code containing path aliases
 * @param {string} url - Absolute path to file containing `code`
 * @param {Options} [options={}] - Resolve options
 * @param {string} [options.tsconfig=pathe.resolve('tsconfig.json')] - Tsconfig
 * @return {string} `code` unmodified or with path aliases resolved
 */
const resolveAliases = (
  code: string,
  url: string,
  options: Options = {}
): string => {
  // set tsconfig path
  options.tsconfig = options.tsconfig ?? pathe.resolve('tsconfig.json')

  /**
   * Tsconfig object.
   *
   * @const {Tsconfig | undefined} tsconfig
   */
  const tsconfig: Tsconfig | undefined = loadTsconfig(
    options.tsconfig,
    options.fileExists,
    options.readFile
  )

  // get compiler options
  const { compilerOptions = {} } = tsconfig ?? {}
  const { baseUrl = '.', paths = {} } = compilerOptions

  // resolve path aliases
  for (const statement of extractStatements(code)) {
    // do nothing if missing module specifier
    if (!statement.specifier) continue

    /**
     * Resolved module specifier.
     *
     * @const {string} specifier
     */
    const specifier: string = resolveAlias(statement.specifier, {
      ...options,
      baseUrl: pathe.resolve(pathe.dirname(options.tsconfig), baseUrl),
      paths,
      tsconfig: undefined,
      url
    })

    // do nothing if specifier did not contain path alias
    if (specifier === statement.specifier) continue

    // replace path alias
    code = code.replace(
      statement.code,
      statement.code.replace(statement.specifier, specifier)
    )
  }

  return code
}

export default resolveAliases
