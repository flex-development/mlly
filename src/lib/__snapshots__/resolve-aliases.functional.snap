// Vitest Snapshot v1

exports[`functional:lib/resolveAliases > should resolve path aliases in code 1`] = `
"/**
 * @file Fixtures - Path Aliases
 * @module fixtures/path-aliases
 */

import '../src/lib/extract-statements'

require('../src/lib/resolve-aliases')
require('./aggregate-error-ponyfill.cjs')

await import('../src/lib/detect-syntax')

export * from '../src/constants'
export * from '../src/interfaces'
"
`;
